maksumov/rss-shop-be
# Task 9 Solution

# Task 9 Definition

Task defined at https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task9-bff/task.md


## Task 9.1

- [x] A working and correct express application should be in the [bff-service folder](https://github.com/maksumov/rss-shop-be/tree/feat/task-9/src/bff-service). Reviewer can start this application locally with any [valid configuration in the .env file](https://github.com/maksumov/rss-shop-be/blob/feat/task-9/src/bff-service/.env.sample) and this application should works as described in the [task 9.](https://github.com/maksumov/rss-shop-be/issues/12)

### How to test Task 9.1.

Clone repo & install dependencies with following commands:
```bash
git clone https://github.com/maksumov/rss-shop-be.git
cd rss-shop-be
git fetch origin pull/14/head:task-9
git checkout task-9
cd src/bff-service
npm install
```

After installing dependencies you should create .env file in the root folder of bff-service (see [.env.sample](https://github.com/maksumov/rss-shop-be/blob/feat/task-9/src/bff-service/.env.sample) for example of configuration or just rename this file to .env file).

And finally you need to compile TypeScript files by running `tsc` and then you can run service by `npm run start:dev`.

This PR closes #12 

## Task 9.2

- [x] The bff-service should be deployed with Elastic Beanstalk. 
- [x] The bff-service call should be redirected to the appropriate service : product-service or CART.
- [x] The response from the bff-service should be the same as if product-service or CART services were called directly.

This PR closes #13 

## Additional Tasks

### Add Cache

- [x] Add a cache at the bff-service level for a request to the getProductsList function of the product-service. 
- [x] The cache should expire in 2 minutes.

#### How to test caching

- Get products list from [BFF link](http://maksumov-bff-api-dev.eu-west-1.elasticbeanstalk.com/products)
- Create new product (see example below): 
```bash
curl --header "Content-Type: application/json" --request POST --data '{"count":20,"price":100,"description":"Это описание нового продукта","title":"Новый продукт"}' https://12yhrnlj11.execute-api.eu-west-1.amazonaws.com/dev/products
```
- Get products list (again from [BFF Link](http://maksumov-bff-api-dev.eu-west-1.elasticbeanstalk.com/products)) - result shouldn’t have new product
- Wait more than 2 minutes
- Get products list from [BFF Link](http://maksumov-bff-api-dev.eu-west-1.elasticbeanstalk.com/products) - result should have new product

#### How to test caching (alternative method)

- Start this application locally (like testing Task 9.1)
- Get products list from local url (like [http://localhost:7000/products](http://localhost:7000/products)) a few times in a row
- Look in the console to find messages how each request was processed: from the cache or using the request (`Request ${url} is handled from cache...` or `Request ${url} is handled from service request...`)
- Wait more than 2 minutes from the first request and make another request for the product list
- Look in the console to find how the request was processed: from the cache or using the request

This PR closes #15 

### NestJS implementation


Working on closing the task #16 

# Self-check

Main tasks score: 5/5
Additional tasks score: 1/2
Total expected score: 6/7

# Data for Task Evaluation

## URLs 

- FE App: 
  - [Store URL](https://d3sod2gr47frq2.cloudfront.net/)
- Product Service: 
  - [Direct link](https://12yhrnlj11.execute-api.eu-west-1.amazonaws.com/dev/products)
  - [BFF Link](http://maksumov-bff-api-dev.eu-west-1.elasticbeanstalk.com/products)
- Cart Service:
  - [Direct Link to Beanstalk URL](http://maksumov-cart-api-dev.eu-west-1.elasticbeanstalk.com/api/profile/cart)
  - [Api Gateway Link to Beanstalk](https://u7dyaoudn4.execute-api.eu-west-1.amazonaws.com/dev/api/profile/cart)
  - [BFF Link](http://maksumov-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart/profile/cart)
